src/v/cluster/rm_stm.cc

```diff
@@ -2665,4 +2665,15 @@ ss::future<> rm_stm::clear_old_tx_pids() {
       });
 }
 
+ss::future<> rm_stm::clear_old_idempotent_pids() {
+    while (_log_state.lru_idempotent_pids.size() > 100500) {
+        auto pid_for_delete = _log_state.lru_idempotent_pids.front().entry.pid;
```

Comment:
I can't validate on the spot if it's safe to remove `pid` while it is being processing (`clear_old_idempotent_pids` may be executed concurrently with `replicate_seq`). We should have per pid locks, take write lock here and read lock in replicate_seq to make sure they are mutually exclusive.